### PR TITLE
Implement add file and folder

### PR DESCRIPTION
**Pull request by matzipan**
Depends on #104 

The lines for `newly_created_path` are currently commented out as they will be implemented in a further pull request.

### CODE REVIEW

File: plugins/folder-manager/File.vala [hunk 1]
```diff
@@ -106,6 +106,17 @@ namespace Scratch.Plugins.FolderManager {
                 return false;
             }
         }
+        
+        // Files can be executed and folders can be cd'd into
+        public bool is_executable {
+            get {
+                try {
+                    return get_boolean_file_attribute(GLib.FileAttribute.ACCESS_CAN_EXECUTE);
```
Comment (jeremypw) [inline plugins/folder-manager/File.vala]: Missing space before (

File: plugins/folder-manager/File.vala [hunk 2]
```diff
@@ -134,6 +145,12 @@ namespace Scratch.Plugins.FolderManager {
             }
         }
         
+        private bool get_boolean_file_attribute(string attribute) throws GLib.Error {
+            var info = file.query_info(attribute, GLib.FileQueryInfoFlags.NONE);
+                
+            return info.get_attribute_boolean(attribute);
```
Comment (jeremypw) [inline plugins/folder-manager/File.vala]: Missing space before (

File: plugins/folder-manager/File.vala [hunk 3]
```diff
@@ -134,6 +145,12 @@ namespace Scratch.Plugins.FolderManager {
             }
         }
         
+        private bool get_boolean_file_attribute(string attribute) throws GLib.Error {
+            var info = file.query_info(attribute, GLib.FileQueryInfoFlags.NONE);
```
Comment (jeremypw) [inline plugins/folder-manager/File.vala]: Missing space before ( 
Previous line: missing space before (

**Comment by jeremypw** [comment]
Blocked pending resolution of conflicts.

File: plugins/folder-manager/FileItem.vala [hunk 4]
```diff
@@ -29,18 +29,16 @@ namespace Scratch.Plugins.FolderManager {
...
             rename_item.activate.connect (() => view.start_editing_item (this));
-
-            var menu = new Gtk.Menu ();
-            menu.append (trash_item);
             menu.append (rename_item);
 
-            menu.show_all ();
+            var delete_item = new Gtk.MenuItem.with_label (_("Move to Trash"));
+            delete_item.activate.connect (() => trash ());
+            menu.append (delete_item);
```
Comment (danrabbit) [inline plugins/folder-manager/FileItem.vala]: Please revert these changes. This makes this file more disorganized than it was before

File: plugins/folder-manager/FolderItem.vala [hunk 5]
```diff
@@ -80,6 +98,18 @@ namespace Scratch.Plugins.FolderManager {
             }
         }
 
+        private void do_close () {
+            /*view.close_folder (path);*/
```
Comment (danrabbit) [inline plugins/folder-manager/FolderItem.vala]: This seems to be a dummy method. Either uncomment this code or remove the method

File: plugins/folder-manager/FolderItem.vala [hunk 6]
```diff
@@ -123,5 +153,57 @@ namespace Scratch.Plugins.FolderManager {
...
+                new_folder = file.file.get_child (_("untitled folder %d").printf (n));
+                n++;
+            }
+
+            try {
+                expanded = true;
+
+                new_folder.make_directory ();
+
+                /*newly_created_path = new_folder.get_path ();*/
```
Comment (danrabbit) [inline plugins/folder-manager/FolderItem.vala]: please remove commented code

File: plugins/folder-manager/FolderItem.vala [hunk 7]
```diff
@@ -123,5 +153,57 @@ namespace Scratch.Plugins.FolderManager {
...
+                new_file = file.file.get_child (_("new file %d").printf (n));
+                n++;
+            }
+
+            try {
+                expanded = true;
+
+                new_file.create (FileCreateFlags.NONE);
+
+                /*newly_created_path = new_file.get_path ();*/
```
Comment (danrabbit) [inline plugins/folder-manager/FolderItem.vala]: please remove commented code

File: plugins/folder-manager/FolderItem.vala [hunk 8]
```diff
@@ -53,16 +53,27 @@ namespace Scratch.Plugins.FolderManager {
         }
 
         public override Gtk.Menu? get_context_menu () {
+            var open_item = new Gtk.MenuItem.with_label (_("Open"));
```
Comment (danrabbit) [inline plugins/folder-manager/FolderItem.vala]: I think we might have talked about this before, but I don't think this function does what it sounds like it does. It sounds like "Open in File Manager", but it's really something like add to sidebar.

I'm not really sure this function is terribly useful, tbh. But if it is, I think it should be renamed

File: plugins/folder-manager/FolderItem.vala [hunk 9]
```diff
@@ -53,16 +53,27 @@ namespace Scratch.Plugins.FolderManager {
...
+
             var rename_item = new Gtk.MenuItem.with_label (_("Rename"));
             rename_item.activate.connect (() => view.start_editing_item (this));
 
-            var trash_item = new Gtk.MenuItem.with_label (_("Move to Trash"));
-            trash_item.activate.connect (trash);
+            var new_file_item = new Gtk.MenuItem.with_label (_("Add File"));
+            new_file_item.activate.connect (() => add_file ());
+
+            var new_folder_item = new Gtk.MenuItem.with_label (_("Add Folder"));
```
Comment (danrabbit) [inline plugins/folder-manager/FolderItem.vala]: I think we should follow Files here and have a "New" submenu with "Folder" and "Empty File".

File: plugins/folder-manager/FolderItem.vala (hunk 8 above)
Comment (matzipan) [inline plugins/folder-manager/FolderItem.vala]: I tried looking for the old conversation today but it's lost in the Slack ether. Will remove.

File: plugins/folder-manager/FolderItem.vala (hunk 9 above)
Comment (matzipan) [inline plugins/folder-manager/FolderItem.vala]: On it
